SHA Based Bundle Digest EP (#107)
operator-framework/enhancements: enhancements/sha-bundle-digest.md

---
title: SHA-Bundle-Digest
authors:
  - '@theishshah'
reviewers:
  - '@camilamacedo86'
  - '@joelanford'
  - '@jmrodri'
  - '@ryantking'
  - '@tlwu2013'
approvers:
  - '@tlwu2013'
  - '@jmrodri'
  - '@joelanford'

creation-date: 2022-01-14

last-updated: 2022-01-14

status: implementable
---

# Support SHA-based Image Tags

## Summary

> Currently, operators generated by `operator-sdk` will store and retrieve container images based on the image name and
> the image tag in the bundle. This EP adds support to `operator-sdk` such that images can be stored and retrieved in the
> bundle with an image name and an image digest as an optional replacement for a tag. Image digests are SHA-256 sums that
> are hashed from the image layer therefore uniquely identifying one specific image.

Reviewer: ```suggestion
Operators built with `operator-sdk` use Docker images that are referenced via an image URL and tag name. This EP adds
the ability to create operators that use image digests instead of tags for operator images as well as related images. An
image digest is a `SHA-256` sum that is created by hashing the contents of a Docker image manifest. A Docker image
manifest is a JSON file that defines the image, most notably its layers. See [Appendix A](##appendix-a) for an example
manifest for `golang:1.17.6`. An alternative method to specify an image is to use its digest directly instead of a tag.
For example, an image for go 1.17.6 would be pulled with the Docker command `docker pull golang:1.17.6`. The exact same
image can be pulled by its digest as well:
`docker pull golang@sha256:197db40b76cce7e70d59e59cc567896395730c6491e7b3da5d1f4c22efa72404`. An important distinction
is that since digests are a hash of the image manifests, any change to the image, will cause it to have a distinct
digest. This means that if a new image is tagged as `golang:1.17.6`, the first `docker pull` command will pull that new
image while the second `docker pull` command will return the same image as before since it uses the digest to reference
it.
```

## Motivation

> Many users run Kubernetes in an air-gapped environment (also known as an offline or disconnected environment), which is
> when a cluster network does not have any access to the public internet, usually for security purposes. In this setup,
> the digest of an image will be different than it is in the public repository since the image location changes. For
> example, if a user imports the image `quay.io/golang:1.17` into their air-gapped registry as
> `quay.internaldomain.net/golang:1.17`, the digest of the image will be different due to the different name even though
> the contents of the images are identical. Using digest SHAs instead of tags in the air-gapped environment will keep the
> digest from changing registry to registry.

Reviewer: ```suggestion
Many users run Kubernetes in an air-gapped environment (also known as an offline or disconnected environment), which is
when a cluster network does not have any access to the public internet, usually for security purposes. In this setup,
since the image is hosted in a registry within the air-gapped environment, it will have a different name. For example,
if you push `golang:1.17.6` to an internal registry, it will have a name along the lines of
`registry.internaldomain.net/library/golang:1.17.6`. Since the name, which is part of the image manifest, is different,
the image digest will be different. This is a problem for security-conscience users who want to ensure that they are
using the exact image they expect since by definition the digests of the same image in two different registries will be
different. This EP resolves that by supporting an operator configuration that works directly with image digests instead
of tags. Users will then be able to push images to their internal registry using the image's digest instead of its tag,
which will keep the digest from changing, and then the operator will pull the image based on its digest.
```

### Goals

> - Give users the ability to enable digest-based image resolution

Reviewer: ```suggestion
- Give users the ability to enable digest-based image resolution.
```

> - Allow users to specify additional images to resolve digests for.

Reviewer: ```suggestion
- Allow users to specify related images to resolve digests for.
```

> - Add ability to resolve image digests from image tag dynamically using local docker registry.

Reviewer: This feels like a proposal. Replacing with different goals that I think are missing.
```suggestion
- Use image digests in the output `ClusterServiceVersion` (see [Appendix B](#appendix-b)).
    - Images specified in deployments will have digests.
    - The `spec.relatedImage` section of the CSV will contain a mapping of images to their digests.
```

### Non-Goals

- Provide an automated way to migrate existing Operator-SDK-based operators from image tags to image digests.
- Identify locations in user code that reference images by tag.

## Proposal

- Add `--use-image-digests` to `operator-sdk generate bundle` command. When enabled, the CLI will:

>   - Search the environment defined in `manager.yaml` for all variables that begin with `RELATED_IMAGE_` to build a list
>     of all images that need digests in the output cluster CSV.
>   - Resolve the digests for the operator image and any additional related images it finds in the environment.
>   - Output the images with digests instead of tags in the cluster CSV.

Reviewer: ```suggestion
  1. Search the environment defined in `manager.yaml` for all variables that begin with `RELATED_IMAGE_` to build a list
    of all related images that need to be added to the CSV (see Appendix C).
  2. Resolve the digests for the operator image and any additional related images it finds in the environment.
  3. Output the operator image(s) with digests instead of tags in the CSV.
  4. Add related images with their digests to the CSV.
```

- Add a variable to the `Makefile` called `USE_IMAGE_DIGESTS`.

> - During bundle generation, discover all environment variables that begin with `RELATED_IMAGE_` then resolve the tags to
>   digests. These variables should be set in `config/manager/manager.yaml` as suggested in
>   [https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions](https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions).
> - Update CSV relatedImages with resolved image tags to digests

Reviewer: Already explained in the first section.
```suggestion
```

> ```yaml
> # Image w/ traditional tags
> - name: RELATED_IMAGE_che_server_secure_exposer_jwt_proxy_image
>   value: quay.io/eclipse/che-jwtproxy:0.10.0
> 
> # Image w/ SHA based digest
> - name: RELATED_IMAGE_single_host_gateway
>   value: quay.io/eclipse/che--traefik:v2.5.0-eb30f9f09a65cee1fab5ef9c64cb4ec91b800dc3fdd738d62a9d4334f0114683
> ```
> 

Reviewer: ```suggestion
```

Moved this to the appendix

  In order to allow users full control over the command's flags, the `Makefile` should use an intermediary variable to
  hold the flags that a user can overwrite entirely. For example:

```Makefile
GENERATE_BUNDLE_FLAGS ?= ""
USE_IMAGE_DIGESTS ?= false
ifeq ($(GENERATE_BUNDLE_FLAGS),"")
    ifeq($(USE_IMAGE_DIGESTS),true)
        GENERATE_BUNDLE_FLAGS += "--use-image-digests"
    endif
endif
bundle:
    operator-sdk generate bundle $(GENERATE_BUNDLE_FLAGS)
```

- The `RELATED_IMAGE_*` variables should be defined in the `manager.yaml` file as outlined
  [here](https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions).
- In order to resolve digests from tags, the CLI will use the `imageresolver` package in
  [operator-manifest-tools](https://github.com/operator-framework/operator-manifest-tools).
- The output cluster CSV will have the `relatedImage` annotation that is explained in the
  [OpenShift Documentation](https://docs.openshift.com/container-platform/4.9/operators/operator_sdk/osdk-generating-csvs.html#olm-enabling-operator-for-restricted-network_osdk-generating-csvs).

### Test Plan

> The resulting artifact of this EP will be a cluster CSV with image digests instead of tags in the image definitions as
> well as digest annotations. This feature will need an integration test that veifiesthat the knobs we provide to enable
> and configure image digests generates an expected CSV. The expected CSV will be based off of the existing memcached CSV,
> located
> [here](https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v3/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml)
> with the `spec.relatedImages` section from the proposal section added to it.

Reviewer: ```suggestion
The resulting artifact of this EP will be a CSV with image digests instead of tags in the image definitions as well as
digest annotations. This feature will need an integration test that verifies that the knobs we provide to enable and
configure image digests generate an expected CSV. The expected CSV will be based on the example memcached CSV with the
proposed changes (see [Appendix B](#appendix-b)).
```

## Implementation History

There has been no implementation to this point.

## Drawbacks

The original motivation for this feature is used in a disconnected environment with an SDK-built operator, OLM, and OCP.
Performing a full E2E test in a complex environment that will need to be specifically built out for this task is out of
scope for this SDK EP.

## Alternatives

We are not considering any alternative solutions at this time.

## Open Questions

- Operator-manifest-tools relies on an external binary, skopio. In order to remove the external binary, can we PR
  operator-manifest-tools to replace the calls to the skopeo CLI with library calls?
- Store digest and tag mappings in file so that user-written code can resolve tags to digest.
- What is the function of `spec.relatedImage.name`?

> - How are the values of `spec.relatedImage.name` used?

Reviewer: ```suggestion
- How are the values of `spec.relatedImage.name` used?

## Appendix A

The following is an example of an image manifest for `golang:1.17.6`:

```json
[
  {
    "Config": "8b86bf336a01235faf28137dae90772076e6f431a2951259d949eb9153012755.json",
    "RepoTags": [
      "golang:1.17.6"
    ],
    "Layers": [
      "644b6e1dc104cf7c01e3d906bad42ca6386e62e03ba55c23c9229eeca2aedf4e/layer.tar",
      "dbaefc91c060fe8e827580d1018824c5479bb8eeadf644b44f9bbdc6cc0ad314/layer.tar",
      "7164fcc370aec09c29fb107f4b404d667a046d7eb652130cfcad3b493d7f9485/layer.tar",
      "ded273d40fe1a66f4cc0e6cc15236a8800c4e465df4df2b682422a5477910951/layer.tar",
      "eac35aa9198a17ca6942cda2efacf8b052fcfd7c9046cace7290a0b80bd43d7b/layer.tar",
      "34cd146535bd6d8585e4bc411593bc82a336a335fbcc0a1cf78b1ecfe635c549/layer.tar",
      "0aafe4d111d6dbd2414e183de18f5736f470de5b0e750b05bbe4063f00366386/layer.tar"
    ]
  }
]

The manifest is built out of a list of layers, repo tags, and a config file. Since the config file and layers themselves
are hashed, if any of them change or the repo tags, the manifest digest will change.

## Appendix B

A `ClusterServiceVersion` is a [CRD](https://olm.operatorframework.io/docs/concepts/crds/clusterserviceversion/) defined by [Operator Lifecycle Manager](https://github.com/operator-framework/operator-lifecycle-manager) that defines metadata about the entire operator.
```

An example CSV for the Memcached example operator is available [here](https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v3/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml).

This EP will change two things on the CSV: use digests in deployment container image locations and add the `spec.relatedImages` section with all other images, such as base images, that need to have a digest resolved from a tag. An abbreviated version of the CSV with the aforementioned changes looks like this:

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: ClusterServiceVersion
metadata:
  # ...
spec:
  # ...
  install:
    spec:
      # ...
      deployments:
      - name: memcached-operator-controller-manager
        spec:
          # ...
          template:
            # ...
            spec:
              containers:
              - args:
                  # ... 
                image: gcr.io/kubebuilder/kube-rbac-proxy@sha256:e10d1d982dd653db74ca87a1d1ad017bc5ef1aeb651bdea089debf16485b080b
                # ...
              - args:
                  # ... 
                image: quay.io/example/memcached-operator@sha256:d044b878fbcbd0c06c2bcc050f3f9ca070ffb089e6bac01accddcd0ba85b7b9e
                # ...
      # ...
    # ...
  # ...
  relatedImages:
  - name: golang:1.17
    image: # TODO: fill out this
  - name: gcr.io/distroless/static:nonroot
    image: # TODO: fill out this
```

## Appendix C

TODO(ryantking): Example and explanation of manager.yaml file. 
```

Author: Added manager yaml sample :+1:

> ---
> 
> title: SHA-Bundle-Digest
> 
> authors:
> 
> - "@theishshah"
> 
> reviewers:
> 
> - TBD
> 
> approvers:
> 
> - TBD
> 
> creation-date: 2022-01-14
> 
> last-updated: 2022-01-14
> 
> status:implementable
> 
> ---

Reviewer 2: Remove the extra white space placed here by the Google doc copy.

```suggestion
---
title: SHA-Bundle-Digest
authors:
- "@theishshah"
reviewers:
- @camilamacedo86 
- @joelanford 
- @jmrodri
- @ryantking 
- @tlwu2013 
approvers:
- @tlwu2013 
- @jmrodri
- @joelanford 
creation-date: 2022-01-14
last-updated: 2022-01-14
status:implementable
---
```

> 
>   
> 
> # Title
> 
>   
> 
> Add support for SHA-based Digest for Images
> 
>   
> 

Reviewer 2: ```suggestion

# Title
Add support for SHA-based Digest for Images

```

> ## Summary
> 
>   
> 
> Currently, operators generated by operator-sdk will store and retrieve container images based on the image name and the image tag in the bundle. This EP adds support to Operator-SDK such that images can be stored and retrieved in the bundle with an image name and an image digest as an optional replacement for a tag. Image digests are SHA-256 sums that are hashed from the image layer therefore uniquely identify one specific image.
> 
>   

Reviewer 2: ```suggestion
## Summary

Currently, operators generated by `operator-sdk` will store and retrieve container images based on the image name and the image tag in the bundle. This EP adds support to `operator-sdk` such that images can be stored and retrieved in the bundle with an image name and an image digest as an optional replacement for a tag. Image digests are SHA-256 sums that are hashed from the image layer therefore uniquely identifying one specific image.
```

> ## Motivation
> 
>   
> 
> This set of changes will allow SDK users in disconnected environments to generate bundles with the image digest
> 
>   
> 
> Users in air gapped environments are currently unable to get images based on tags, and can only access local registries based on the digest
> 
> This change is necessary due to non-deterministic behavior with images identified via tags, this issue specifically occurs in the OCP registry due to how the digest is generated with tag name as inputs
> 
>   

Reviewer 2: Again the whitespace from the Google doc makes it distracting. And I moved a few sentences around.

```suggestion
## Motivation

Users in air-gapped (disconnected) environments are currently unable to get images based on tags, and can only access local registries based on the digest.

This set of changes will allow SDK users in air-gapped (disconnected) environments to generate bundles with the image digest. This change is necessary due to non-deterministic behavior with images identified via tags, this issue specifically occurs in the OCP registry due to how the digest is generated with tag name as inputs.
```

> ### Goals
> 
>   
>   
> 
> -   Give users the ability to the enable digest-based image resolution via a CLI flag when generating bundles.
>     
> -   Allow user to specify additional images to resolve digests via an environment variable.
>     
> -   Add ability to resolve image digests from image tag dynamically using local docker registry.
>     
> -   OPEN QUESTION: Store digest and tag mappings in file so that user-written code can resolve tags to digest.
>     
> 
>   

Reviewer 2: ```suggestion
### Goals

-   Give users the ability to the enable digest-based image resolution via a CLI flag when generating bundles.
-   Allow users to specify additional images to resolve digests via an environment variable.
-   Add the ability to resolve image digests from image tag dynamically using local docker registry.
-   OPEN QUESTION: Store digest and tag mappings in file so that user-written code can resolve tags to digest.
```

> ### Non-Goals
> 
>   
> 
> -   Migrate existing Operator-SDK-based operators from image tags to image digests.
>     
> -   Identify locations in user code that reference images by tag.
>     
> 
>   
>   

Reviewer 2: ```suggestion
### Non-Goals

-   Migrate existing `operator-sdk` based operators from image tags to image digests.
-   Identify locations in user code that reference images by tag.
```

Reviewer 2: Also add a reference to the Drawback mentioned below. That is also a NON goal. You could just reference it as a link 
```
- [See Drawbacks section](#drawbacks)

## Proposal

> -   Add `--image-digests` flag to `make bundle` in order to enable this feature to be run. TODO(ish): Add example command being run with this flag.

Reviewer 2: Is this flag added to the `operator-sdk generate bundle` command that will be called by the `bundle` make target? OR is this actually suggesting we're adding a flag to the `make bundle` target? I think the example will likely answer this question but wanted to make sure it was asked.

Author: The flag is added to the SDK command which is called by the Makefile. The change in the Makefile will be to have a user toggle a variable to enable or disable running the generate bundle command with this flag (see Design Details)

> ## Proposal
> 
> -   Add `--image-digests` flag to `make bundle` in order to enable this feature to be run. TODO(ish): Add example command being run with this flag.
>     
> -   During bundle generation, discover all environment variables that begin with `RELATED_IMAGE_` then resolve the tags to digests. These variables should be set in `config/manager/manager.yaml` as suggested in [https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions](https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions). 
>     
> 
> ```
>  - name: RELATED_IMAGE_che_server_secure_exposer_jwt_proxy_image
> 
>    value: quay.io/eclipse/che-jwtproxy:0.10.0
> 
> - name: RELATED_IMAGE_single_host_gateway
> 
>   value: quay.io/eclipse/che--traefik:v2.5.0-eb30f9f09a65cee1fab5ef9c64cb4ec91b800dc3fdd738d62a9d4334f0114683
> ```
> 
> -   Generate OCP cluster CSV to allow the docker daemon to resolve image tags to digests. 
>     
> 
>   
>   

Reviewer 2: Just whitespace issues here.

```suggestion
## Proposal

-   Add `--image-digests` flag to `make bundle` in order to enable this feature to be run. TODO(ish): Add example command being run with this flag.

-   During bundle generation, discover all environment variables that begin with `RELATED_IMAGE_` then resolve the tags to digests. These variables should be set in `config/manager/manager.yaml` as suggested in [https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions](https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions). 

```
 - name: RELATED_IMAGE_che_server_secure_exposer_jwt_proxy_image
   value: quay.io/eclipse/che-jwtproxy:0.10.0
- name: RELATED_IMAGE_single_host_gateway
  value: quay.io/eclipse/che--traefik:v2.5.0-eb30f9f09a65cee1fab5ef9c64cb4ec91b800dc3fdd738d62a9d4334f0114683
```

-   Generate OCP cluster CSV to allow the docker daemon to resolve image tags to digests. 
```

Reviewer 2: I don't see anything stating that we will generate the `CSV.spec.relatedImages` field. Are we not doing that? @tlwu2013

Reviewer 3: > spec.relatedImages (user): a list of image tags containing SHA digests mapped to in-CSV names that your Operator might require to perform their functions. 

I'd vote for generating the`CSV.spec.relatedImages` field (see [in the doc](https://sdk.operatorframework.io/docs/olm-integration/generation/#csv-fields)) since it requires SHA digests too. This would help a lot rather than having users fill that manually.

BTW, regarding those example links, it'll be better to include the links to the original file so readers can easily see the raw file to get a better idea.

e.g. Taking a community Operator, eclipse-che Operator, as an example:

**1.** Author needs to add `RELATED_IMAGE_*` in `config/manager/manager.yaml` ([see example](https://github.com/eclipse-che/che-operator/blob/main/config/manager/manager.yaml#L86-L87)).

**2.** `make bundle --image-digests` generates CSV as:
-->
**2-1.** generate SHA ref. in CSV's deployment's env var ([see in example CSV yaml](https://github.com/eclipse-che/che-operator/blob/main/bundle/stable/eclipse-che-preview-openshift/manifests/che-operator.clusterserviceversion.yaml#L1125-L1126)) 

**2-2.** generate CSV's `spec.relatedImage` from deployment's env var ([see in example CSV yaml](https://github.com/eclipse-che/che-operator/blob/main/bundle/stable/eclipse-che-preview-openshift/manifests/che-operator.clusterserviceversion.yaml#L1604-L1606))

> ### Risks and Mitigations
> 
>   
> 
> Since the feature is op-in, it carries the risk of the end-user enabling it without a correct understanding of the feature and its implications. This can be mitigated with clear documentation both on the website and in the CLI help section.
> 
>   
>   

Reviewer 2: ```suggestion
### Risks and Mitigations

Since the feature is opt-in, it carries the risk of the end-user enabling it without a correct understanding of the feature and its implications. This can be mitigated with clear documentation both on the website and in the CLI help section.
```

> ## Design Details
> 
>   
> 
>  -   Add a `--image-digests` flag to the `operator-sdk generate bundle` command that will enable this functionality.
>     
>  -   Add a variable to the generated Makefile called `USE_IMAGE_DIGESTS` that defaults to false and when enabled, it passes the `--use-image-digests` flag to `operator-sdk generate bundle`.
>     
>  -   When `operator-sdk generate-bundle` is called with the `--use-image-digests` flag enabled, it will execute the following logic.
>     
> 	- Search the environment that is defined in the manager.yaml fi for all variables that begin with `RELATED_IMAGE_` to    build a list of images that need to have tags resolved to digests    along with the Operator’s main image and rbac-proxy image.
>    
> 	   -  For each image on the list, it will use `operator-manifest-tools` to resolve the image tag to a digest
>    
> 	   -  When building the cluster CSV, all images that have been identified as needing a digest instead of a tag should have that  digest as an annotation in the CSV. TODO(ish): add before/after    snippet from the CSV to show the difference.
>     
>  

Reviewer 2: Whitespace and some formatting.

```suggestion
## Design Details

 -   Add a `--image-digests` flag to the `operator-sdk generate bundle` command that will enable this functionality.
 -   Add a variable to the generated `Makefile` called `USE_IMAGE_DIGESTS` that defaults to `false` and when enabled, it passes the `--use-image-digests` flag to `operator-sdk generate bundle`.
 -   When `operator-sdk generate-bundle` is called with the `--use-image-digests` flag enabled, it will execute the following logic.
 	1. Search the environment that is defined in the `manager.yaml` file for all variables that begin with `RELATED_IMAGE_` to build a list of images that need to have tags resolved to digests along with the Operator’s main image and rbac-proxy image.
 	 1.  For each image in the list, it will use `operator-manifest-tools` to resolve the image tag to a digest
 	 1.  When building the cluster CSV, all images that have been identified as needing a digest instead of a tag should have that  digest as an annotation in the CSV. TODO(ish): add before/after snippet from the CSV to show the difference.
```

> ### Test Plan
> 
>   
> 
> The resulting artifact of this EP will be a cluster CSV with image digests instead of tags in the image definitions as well as digest annotations. This feature will need an integration test that veifiesthat the knobs we provide to enable and configure image digests generates an expected CSV. The expected CSV will be manually created based on the expectations for how the feature should work. We can use this CSV in existing Operator Registry testing setups to verify that OPM correctly renders it.
> 
>   

Reviewer 2: whitespace and one typo.

```suggestion
### Test Plan

The resulting artifact of this EP will be a cluster CSV with image digests instead of tags in the image definitions as well as digest annotations. This feature will need an integration test that veifies that the knobs we provide to enable and configure image digests generates an expected CSV. The expected CSV will be manually created based on the expectations for how the feature should work. We can use this CSV in existing Operator Registry testing setups to verify that OPM correctly renders it.
```

> ## Implementation History
> 
>   
> 
> There has been no implementation to this point.
> 
>   

Reviewer 2: ```suggestion
## Implementation History

There has been no implementation to this point.
```

> ## Drawbacks
> 
>   
> 
> The original motivation for this feature is used in a disconnected environment with an SDK-built operator, OLM, and OCP. Performing a full E2E test in a complex environment that will need to be specifically built out for this task is out of scope for this SDK EP.
> 

Reviewer 2: ```suggestion
## Drawbacks

The original motivation for this feature is used in a disconnected environment with an SDK-built operator, OLM, and OCP. Performing a full E2E test in a complex environment that will need to be specifically built out for this task is out of scope for this SDK EP.
```

> ## Alternatives
> 
>   
> 
> We are not considering any alternative solutions at this time.

Reviewer 2: ```suggestion
## Alternatives

We are not considering any alternative solutions at this time.
```

- The output cluster CSV will have the `relatedImage` annotation that is explained in the
  [OpenShift Documentation](https://docs.openshift.com/container-platform/4.9/operators/operator_sdk/osdk-generating-csvs.html#olm-enabling-operator-for-restricted-network_osdk-generating-csvs).

### Test Plan

The resulting artifact of this EP will be a CSV with image digests instead of tags in the image definitions as well as
digest annotations. This feature will need an integration test that verifies that the knobs we provide to enable and
configure image digests generate an expected CSV. The expected CSV will be based on the example memcached CSV with the
proposed changes (see [Appendix B](#appendix-b)).

## Implementation History

There has been no implementation to this point.

## Drawbacks

The original motivation for this feature is used in a disconnected environment with an SDK-built operator, OLM, and OCP.
Performing a full E2E test in a complex environment that will need to be specifically built out for this task is out of
scope for this SDK EP.

## Alternatives

We are not considering any alternative solutions at this time.

## Open Questions

> - Operator-manifest-tools relies on an external binary, skopio. In order to remove the external binary, can we PR

Reviewer 2: ```suggestion
- `operator-manifest-tools` relies on an external binary, `skopeo`. In order to remove the external binary, can we PR
```

  operator-manifest-tools to replace the calls to the skopeo CLI with library calls?

>   - Non-blocking question, a future improvment will be to add library support instead of using an external binary.

Reviewer 2: ```suggestion
  - Non-blocking question, a future improvement will be to add library support instead of using an external binary.
```

    "RepoTags": ["golang:1.17.6"],
    "Layers": [
      "644b6e1dc104cf7c01e3d906bad42ca6386e62e03ba55c23c9229eeca2aedf4e/layer.tar",
      "dbaefc91c060fe8e827580d1018824c5479bb8eeadf644b44f9bbdc6cc0ad314/layer.tar",
      "7164fcc370aec09c29fb107f4b404d667a046d7eb652130cfcad3b493d7f9485/layer.tar",
      "ded273d40fe1a66f4cc0e6cc15236a8800c4e465df4df2b682422a5477910951/layer.tar",
      "eac35aa9198a17ca6942cda2efacf8b052fcfd7c9046cace7290a0b80bd43d7b/layer.tar",
      "34cd146535bd6d8585e4bc411593bc82a336a335fbcc0a1cf78b1ecfe635c549/layer.tar",
      "0aafe4d111d6dbd2414e183de18f5736f470de5b0e750b05bbe4063f00366386/layer.tar"
    ]
  }
]
```

The manifest is built out of a list of layers, repo tags, and a config file. Since the config file and layers themselves
are hashed, if any of them change or the repo tags, the manifest digest will change.

## Appendix B

A `ClusterServiceVersion` is a [CRD](https://olm.operatorframework.io/docs/concepts/crds/clusterserviceversion/) defined by [Operator Lifecycle Manager](https://github.com/operator-framework/operator-lifecycle-manager) that defines metadata about the entire operator.

> ````

Reviewer 2: This should either be removed or made into 3.

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: ClusterServiceVersion
metadata:
  # ...
spec:
  # ...
  install:
    spec:
      # ...
      deployments:
        - name: memcached-operator-controller-manager
          spec:
            # ...
            template:
              # ...
              spec:
                containers:
                  - args:
                    # ...
                    image: gcr.io/kubebuilder/kube-rbac-proxy@sha256:e10d1d982dd653db74ca87a1d1ad017bc5ef1aeb651bdea089debf16485b080b
                    # ...
                  - args:
                    # ...
                    image: quay.io/example/memcached-operator@sha256:d044b878fbcbd0c06c2bcc050f3f9ca070ffb089e6bac01accddcd0ba85b7b9e
                    # ...
      # ...
    # ...
  # ...
  relatedImages:
    - name: golang:1.17
      image: golang:8d717e8a7fa8035f5cfdcdc86811ffd53b7bb17542f419f2a121c4c7533d29ee
    - name: gcr.io/distroless/static:nonroot
      image: gcr.io/distroless/static:741704ac0e5e6ff758a8c46f0f028375252fdf147248d95514ce11ff57fdece9

> ````

Reviewer 2: This should be 3 backticks

---
title: SHA-Bundle-Digest
authors:
- "@theishshah"
reviewers:
- @camilamacedo86 
- @joelanford 
- @jmrodri
- @ryantking 
- @tlwu2013 
approvers:
- @tlwu2013 
- @jmrodri
- @joelanford 
creation-date: 2022-01-14
last-updated: 2022-01-14
status:implementable
---

> # Title
> Add support for SHA-based Digest for Images

Reviewer: ```suggestion
# Support SHA-based Digest for Images
```

Think this makes more sense?

## Summary

> Currently, operators generated by `operator-sdk` will store and retrieve container images based on the image name and the image tag in the bundle. This EP adds support to `operator-sdk` such that images can be stored and retrieved in the bundle with an image name and an image digest as an optional replacement for a tag. Image digests are SHA-256 sums that are hashed from the image layer therefore uniquely identifying one specific image.

Reviewer: Supreme nit, but can you blast this doc with prettier? Would be nice to have consistent formatting, especially line widths just to make reading easier and remove the need to review whitespace/fromatting things.

```
$ prettier --parser markdown --print-width 120 --single-quote --write enhancements/sha-bundle-digest.md
```

That should do the trick.

## Motivation

> Users in air-gapped (disconnected) environments are currently unable to get images based on tags, and can only access local registries based on the digest.

Reviewer: I think we should start by defining what air-gapped or disconnected means then using one term the rest of the document.

E.g.
```
Many users run Kubernetes in an air-gapped environment (also known as an offline or disconnected environment), which is when a cluster network does not have any access to the public internet, usually for security purposes. In this setup, the docker daemon is unable to fetch images based on tag because ...
```

After reading this section I should know:
a) What is an air-gapped environment?
b) Why can I not use traditional image tags in an air-gapped environment?
c) How does adding support for digest-based images fix the issue in b?

Currently, this section explains the "what," but it also needs to explain the "why."

Reviewer 2: +1 to all of the above

This set of changes will allow SDK users in air-gapped (disconnected) environments to generate bundles with the image digest. This change is necessary due to non-deterministic behavior with images identified via tags, this issue specifically occurs in the OCP registry due to how the digest is generated with tag name as inputs.

### Goals

> -   Give users the ability to the enable digest-based image resolution via a CLI flag when generating bundles.

Reviewer: The CLI flag is a technical detail that should be in the proposal section. Keep goals as general as possible instead of coupling them to specific implementation details. For example, the goal of adding a toggle to enable digest-based image is something we definitely want, but we might want an environment variable instead of a flag. We don't want discussions on implementation details to muddy our actual goals.

> -   Allow users to specify additional images to resolve digests via an environment variable.

Reviewer: Same comment as above, remove the implementation detail.

-   Add the ability to resolve image digests from image tag dynamically using local docker registry.

> -   OPEN QUESTION: Store digest and tag mappings in file so that user-written code can resolve tags to digest.

Reviewer: Move to the open questions section. I think this ties into our ongoing discussion of the cluster CSV annotations.

### Non-Goals

> -   Migrate existing `operator-sdk` based operators from image tags to image digests.

Reviewer: To be more specific, we are not aiming to provide an automated way to do this.

-   Identify locations in user code that reference images by tag.

## Proposal

> -   Add `--image-digests` flag to `make bundle` in order to enable this feature to be run. 

Reviewer: I thought we opted for an environment variable that will add this flag to the underlying CLI call, not a flag to the Make target itself?

> -   During bundle generation, discover all environment variables that begin with `RELATED_IMAGE_` then resolve the tags to digests. These variables should be set in `config/manager/manager.yaml` as suggested in [https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions](https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions). 

Reviewer: I think we are missing a bullet on the CLI flag for the `operator-sdk generate bundle` command. I think the general flow of the proposal should be:

```
- Add `--use-image-digests` to `operator-sdk generate bundle` command. When enabled, the CLI will:
  - Search the environment defined in `manager.yaml` for all variables that begin with `RELATED_IMAGE_` to build a list of all images that need digests in the output cluster CSV.
  - Resolve the digests for the operator image and any additional related images it finds in the environment. 
  - Output the images with digests instead of tags in the cluster CSV.
- Add a variable to the `Makefile` called `USE_IMAGE_DIGESTS`.
```

Now the section is focused on the specific changes to how the user interacts with the software and how its functionality changes based on that user interaction.

Since the feature is opt-in, it carries the risk of the end-user enabling it without a correct understanding of the feature and its implications. This can be mitigated with clear documentation both on the website and in the CLI help section.

## Design Details

 -   Add a `--image-digests` flag to the `operator-sdk generate bundle` command that will enable this functionality.
 -   Add a variable to the generated `Makefile` called `USE_IMAGE_DIGESTS` that defaults to `false` and when enabled, it passes the `--use-image-digests` flag to `operator-sdk generate bundle`.
 -   When `operator-sdk generate-bundle` is called with the `--use-image-digests` flag enabled, it will execute the following logic.
 	1. Search the environment that is defined in the `manager.yaml` file for all variables that begin with `RELATED_IMAGE_` to build a list of images that need to have tags resolved to digests along with the Operator’s main image and rbac-proxy image.
 	 1.  For each image in the list, it will use `operator-manifest-tools` [https://github.com/operator-framework/operator-manifest-tools](https://github.com/operator-framework/operator-manifest-tools) to resolve the image tag to a digest

>  	 1.  When building the cluster CSV, all images that have been identified as needing a digest instead of a tag should have that  digest as an annotation in the CSV.

Reviewer: This section should focus on the implementation details that the engineer will have to take into account when building out a feature that does what is defined in the proposal. For example:

```
- As of this EP, the new `--use-image-digests` is the only flag to pass to the `operator-sdk generate bundle` command. In order to allow users full control over the command's flags, the `Makefile` should use an intermediary variable to hold the flags that a user can overwrite entirely. For example:
\```
GENERATE_BUNDLE_FLAGS ?= ""
USE_IMAGE_DIGESTS ?= false
ifeq ($(GENERATE_BUNDLE_FLAGS),"")
    ifeq($(USE_IMAGE_DIGESTS),true)
        GENERATE_BUNDLE_FLAGS += "--use-image-digests"
    endif
endif
bundle:
    operator-sdk generate bundle $(GENERATE_BUNDLE_FLAGS)
\```
- The `RELATED_IMAGE_*` variables should be defined in the `manager.yaml` file as outlined [here](https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions).
- In order to resolve digests from tags, the CLI will use the `imageresolver` package in [operator-manifest-tools](https://github.com/operator-framework/operator-manifest-tools).
- The output cluster CSV will have the `relatedImage` annotation that is explained in the [OpenShift Documentation](https://docs.openshift.com/container-platform/4.9/operators/operator_sdk/osdk-generating-csvs.html#olm-enabling-operator-for-restricted-network_osdk-generating-csvs).
```

Now the details directly tie into specific proposal items, which themselves tie into specific goals.
Also the Make code is just written by hand now, verify it if you opt to add it into this proposal.

### Test Plan

> The resulting artifact of this EP will be a cluster CSV with image digests instead of tags in the image definitions as well as digest annotations. This feature will need an integration test that veifies that the knobs we provide to enable and configure image digests generates an expected CSV. The expected CSV will be manually created based on the expectations for how the feature should work. We can use this CSV in existing Operator Registry testing setups to verify that OPM correctly renders it.

Reviewer: Can we put an example CSV for the Memcached demo operator here? It will help with visualization. Then the engineer who implements will know exactly what the output should look like.
